c/src/ml-api-remote-service.c

```diff
@@ -0,0 +1,389 @@
+_mlrs_get_edge_info (ml_option_h option, edge_info_s * edge_info)
+{
+  void *value;
+
+  if (ML_ERROR_NONE == ml_option_get (option, "host", &value))
+    edge_info->host = g_strdup (value);
+  if (ML_ERROR_NONE == ml_option_get (option, "port", &value))
+    edge_info->port = *((guint *) value);
+  if (ML_ERROR_NONE == ml_option_get (option, "dest-host", &value))
+    edge_info->dest_host = g_strdup (value);
+  if (ML_ERROR_NONE == ml_option_get (option, "dest-port", &value))
+    edge_info->dest_port = *((guint *) value);
+  if (ML_ERROR_NONE == ml_option_get (option, "connect-type", &value))
+    edge_info->conn_type = _mlrs_get_conn_type (value);
+  if (ML_ERROR_NONE == ml_option_get (option, "topic", &value))
+    edge_info->topic = g_strdup (value);
+  if (ML_ERROR_NONE == ml_option_get (option, "node-type", &value))
+    edge_info->node_type = _mlrs_get_node_type (value);
+}
+
+/**
+ * @brief Set nns-edge info.
+ */
+static void
+_mlrs_set_edge_info (edge_info_s * edge_info, nns_edge_h edge_h)
+{
+  char port[6];
+
+  nns_edge_set_info (edge_h, "HOST", edge_info->host);
+  sprintf (port, "%u", edge_info->port);
+  nns_edge_set_info (edge_h, "PORT", port);
+
+  if (edge_info->topic)
+    nns_edge_set_info (edge_h, "TOPIC", edge_info->topic);
+
+  nns_edge_set_info (edge_h, "DEST_HOST", edge_info->dest_host);
+  sprintf (port, "%u", edge_info->dest_port);
+  nns_edge_set_info (edge_h, "DEST_PORT", port);
+}
+
+/**
+ * @brief Release edge info.
+ */
+static void
+_mlrs_release_edge_info (edge_info_s * edge_info)
+{
+  g_free (edge_info->dest_host);
+  g_free (edge_info->host);
+  g_free (edge_info->topic);
+}
+
+/**
+ * @brief Get ml remote service type from ml_option.
+ */
+static ml_remote_service_type_e
+_mlrs_get_service_type (gchar * service_str)
+{
+  ml_remote_service_type_e service_type = ML_REMOTE_SERVICE_TYPE_UNKNOWN;
+
+  if (g_ascii_strcasecmp (service_str, "model_raw") == 0) {
+    service_type = ML_REMOTE_SERVICE_TYPE_MODEL_RAW;
+  } else if (g_ascii_strcasecmp (service_str, "model_url") == 0) {
+    service_type = ML_REMOTE_SERVICE_TYPE_MODEL_URL;
+  } else if (g_ascii_strcasecmp (service_str, "pipeline_raw") == 0) {
+    service_type = ML_REMOTE_SERVICE_TYPE_PIPELINE_RAW;
+  } else if (g_ascii_strcasecmp (service_str, "pipeline_url") == 0) {
+    service_type = ML_REMOTE_SERVICE_TYPE_PIPELINE_URL;
+  } else {
+    _ml_error_report ("Invalid service type: %s, Please check service type.",
+        service_str);
+    service_type = ML_REMOTE_SERVICE_TYPE_UNKNOWN;
+  }
+  return service_type;
+}
+
+/**
+ * @brief Process ml remote service
+ */
+static void
+_mlrs_process_remote_service (nns_edge_data_h data_h)
+{
+  void *data;
+  nns_size_t data_len;
+  gchar *service_str = NULL;
+  gchar *service_key = NULL;
+  ml_remote_service_type_e service_type;
+
+  nns_edge_data_get (data_h, 0, &data, &data_len);
+
+  nns_edge_data_get_info (data_h, "service-type", &service_str);
+  service_type = _mlrs_get_service_type (service_str);
```

Comment:
Is```service_str``` vaild always?
How about checking NULL of service_str in ```_mlrs_get_service_type()``` ? 
```

Reply:
I added to check NULL. Thanks.